clang/lib/Sema/SemaOverload.cpp

```diff
@@ -4393,6 +4401,24 @@ getFixedEnumPromtion(Sema &S, const StandardConversionSequence &SCS) {
   return FixedEnumPromotion::ToPromotedUnderlyingType;
 }
 
+static ImplicitConversionSequence::CompareKind
+HLSLCompareFloatingRank(QualType LHS, QualType RHS) {
+  assert(LHS->isVectorType() == RHS->isVectorType() &&
```

Comment:
Is this true? are there no vector splat cases we need to consider where we would promote a scalar to a vector?

Reply:
I'm not yet handling scalar -> vector conversions. That won't change this code, but it will add extra handling in the `CompareStandardConversionSequences` function because we'll need to prefer scalar->vector conversions when the element type is the same (and maybe a promotion).

We don't yet generate the cast sequences for the scalar->vector conversions, so we can't yet handle them.

For context what this is doing: during overload resolution a list of candidate functions is generated by iterating over each possible function and generating possible conversion sequences for each overload. Then each conversion sequence is evaluated against the other to determine if it is better or worse to choose which overload we want. This code just extends that to allow floating point promotions and conversions more freely than C does, and uses the same logic for vector element conversions.